### ISSUE TITLE
Inconsistency when using Object Destructuring and null check

### ISSUE THREAD

**Issue by GabrielCR99** [dart-lang/linter#4889]
**Describe the issue**
When using object destructuring in Dart and forcing it with the null check operator '!', linter warns us that the operator is unnecessary, but when removing it from the code, it gives us the error [pattern_type_mismatch_in_irrefutable_context](https://dart.dev/diagnostics/pattern_type_mismatch_in_irrefutable_context).

**To Reproduce**
Example using dio package
```dart
      final Response(data: {'response': String message}) =
          await Dio().post<Map<String, dynamic>>(
        'https://example.com/login',
        data: {'user': 'foo', 'password': 'bar'},
      );

      final Response(data: {'response': String message}!) =
          await Dio().post<Map<String, dynamic>>(
        'https://example.com/login',
        data: {'user': 'foo', 'password': 'bar'},
      );
```

**Expected behavior**
Probably linter shouldn't complain about using the bang operator '!'

**Additional context**
![image](https://github.com/dart-lang/linter/assets/42328188/592b9aae-2640-428d-8cef-89789acf1806)
![image](https://github.com/dart-lang/linter/assets/42328188/ef411295-e0f7-46d7-b147-915f04d77b58)

**Comment by srawlins** [comment]
Yikes, yes I agree the problem probably lies in the `unnecessary_null_checks` rule, or in some utility providing logic to that rule...

**Comment by GabrielCR99** [comment]
Yeah, looks likes it's indeed! I thought it was something related with configuring analysis_options.yaml file in the language config, but disabling it didn't help unfortunately, those 3:

```yaml
  language:
    # strict-casts: true
    # strict-inference: true
    # strict-raw-types: true
```
